### PR TITLE
Ability to suppress warning "Root (sudo) is required" in plugins [needs revision]

### PR DESCRIPTION
**Pull request by xgin**
Some plugins do not require root privileges.
The warning is misleading, e.g. https://github.com/plesk/letsencrypt-plesk/issues/35
Proposed solution is to expose is_root_required property in order to check root privileges on demand.

### CODE REVIEW

**Comment by bmw** [comment]
Thanks for the PR @xgin. I know these are the changes proposed by @pde but I personally do not like this approach. The reason is that this changes the plugin API and will break plugins that don't have a `is_root_required` property (there is no requirement that they inherit from `letsencrypt.plugins.common.Plugin`).

I'm still in favor of the solution I discussed in IRC of simply removing the warning. To post my reasoning here, the "Root (sudo) is required to run most of letsencrypt functionality" message is only a warning. It does not stop execution so plugins still have to handle the case where they have insufficient permissions. Removing the warning stops suggesting people run the client with root privileges when it may not be necessary (such as when using `manual`, `webroot`, or `plesk` plugins). This approach also goes with the idea that it's easier to ask for forgiveness than permission.

If we did this, I think we should make sure sane error messages are thrown when we have insufficient permissions, but I suspect this will be a pretty straight forward fix.

If pde wants to use this approach, he's certainly able to merge this PR, but I personally think we can do better.

**Comment by xgin** [comment]
@bmw Good point, I have not thought plugins could not be inherited from the `Plugin` class.
We may check the property exists  (`hasattr` or even `try-except`), otherwise treat it as either True or False.
That will not break backward compatibility.

I also agree that drop of the code is always a good solution :)
BTW, I have checked the error messages in plugins:
* standalone
```
Could not bind TCP port 443 because you don't have the appropriate permissions (for example, you aren't running this program as root).
```
* webroot
```
The webroot plugin is not working; there may be problems with your existing configuration.
The error was: PluginError('/root/test does not exist or is not a directory',)
```
```
The webroot plugin is not working; there may be problems with your existing configuration.
The error was: PluginError("Couldn't create root for {0} http-01 challenge responses: {1}", 'ek-local.ru', OSError(13, 'Permission denied'))
```
* apache
```
The selected plugin encountered an error while parsing your server configuration
and cannot be used. The error was:

Error while running apache2ctl configtest.
Action 'configtest' failed.
The Apache error log may have more information.

apache2: Syntax error on line 216 of /etc/apache2/apache2.conf: Could not open
configuration file /etc/apache2/conf-enabled/zz010_psa_httpd.conf: Permission
denied
```

**Comment by pde** [comment]
To be clear, I wasn't suggesting a breaking change to the plugin API for this -- either a warning printed by the plugins that need root themselves, or an API attribute if but only if we can make that property optional for the plugin to implement.

**Comment by xgin** [comment]
@pde, Sure. The interface is not even touched.
Plugins inherited `letsencrypt.plugins.common.Plugin` respond to `is_root_required` with True by default.
I've added case with plugin having no attribute for complete backward compatibility.

File: letsencrypt/cli.py [hunk 1]
```diff
@@ -542,6 +547,19 @@ def record_chosen_plugins(config, plugins, auth, inst):
     cn.installer = plugins.find_init(inst).name if inst else "none"
 
 
+def check_root_privileges(plugin):
+    """Warns user about insufficient privileges."""
+    if not hasattr(plugin, 'is_root_required'):
+        return
+    if not plugin.is_root_required:
+        return
+    if not os.geteuid() == 0:
```
Comment (kuba) [inline letsencrypt/cli.py]: `x != 0` instead of `not x == 0`

File: letsencrypt/cli.py [hunk 2]
```diff
@@ -542,6 +547,19 @@ def record_chosen_plugins(config, plugins, auth, inst):
     cn.installer = plugins.find_init(inst).name if inst else "none"
 
 
+def check_root_privileges(plugin):
+    """Warns user about insufficient privileges."""
+    if not hasattr(plugin, 'is_root_required'):
+        return
+    if not plugin.is_root_required:
+        return
```
Comment (kuba) [inline letsencrypt/cli.py]: it' be a lot easier to just use `getattr(plugin, 'is_root_required', False)`

File: letsencrypt/plugins/common.py [hunk 3]
```diff
@@ -90,6 +90,16 @@ def add_parser_arguments(cls, add):
 
         """
 
+    @property
```
Comment (kuba) [inline letsencrypt/plugins/common.py]: Why make a property of such uncomplicated thing? Just say `is_root_required = True`

**Comment by kuba** [comment]
That PR is very hackish :(

`interfaces.py` should be updated accordingly

**Comment by kuba** [comment]
That being said, I would find it a lot better to just remove the warning as `bmw` says. Those hacks are not worth it.

**Comment by pde** [comment]
OK, I'm going to try to take some time to figure out whether we want to clean up this PR, or take another simpler but (perhaps) less user-helpful route.

**Comment by bmw** [comment]
Despite this change being made in conjunction with our largest 3rd party plugin, I think changing the plugin interface is a bad idea and unnecessary. With this in mind, I'm closing this PR since #1962 has been merged. If you'd like me to reconsider, please let me know!

**Pull request closed without merging**